### PR TITLE
Task/code cleanup

### PR DESCRIPTION
**Pull request by tsteuwer**
A little cleanup of code.

### CODE REVIEW

File: lib/headers.js [hunk 1]
```diff
@@ -2,15 +2,18 @@
 var is = require('type-is');
 
 module.exports = function (app, options) {
+  var remotes = app.remotes();
```
Comment (digitalsadhu) [inline lib/headers.js]: You have broken out remotes here, but up further you used app.remotes() on the same line.

File: lib/deserialize.js [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
-  //can work with.
-
-  remotes.before('**', function (ctx, next) {
-
+  /**
+   * Register a handler to run before all remote methods so that we can
+   * transform JSON API structured JSON payload into something loopback
+   * can work with.
+   */
+  app.remotes().before('**', function (ctx, next) {
```
Comment (digitalsadhu) [inline lib/deserialize.js]: Below you have broken out app.remotes() to var remotes = app.remotes()

**Comment by digitalsadhu** [comment]
Looking nice, tests passing except for the same 1 as before so looks like you didn't break anything

File: .vimrc [hunk 3]
```diff
@@ -0,0 +1,4 @@
+set expandtab
+set tabstop=2
+set shiftwidth=2 
+set softtabstop=2
```
Comment (digitalsadhu) [inline .vimrc]: A vim user!

File: lib/deserialize.js [hunk 4]
```diff
@@ -22,28 +22,56 @@ module.exports = function (app) {
     });
 
     if (matches.length > 0) {
-
-      //set the JSON API Content-Type response header
+      // set the JSON API Content-Type response header
       ctx.res.set({'Content-Type': 'application/vnd.api+json'});
```
Comment (digitalsadhu) [inline lib/deserialize.js]: A side note, we should be able to remove these Content-Type header sets soon as LB should now support switching on the Accept header. 

I plan on expanding loopbacks native support for handling the json api content type soon as well

File: lib/deserialize.js [hunk 5]
```diff
@@ -22,28 +22,56 @@ module.exports = function (app) {
...
-      if (!data) return next();
+      if (!data) {
+        return next();
+      }
 
-      if (!data.data) return next(new Error('JSON API resource object must contain `data` property'));
-      if (!data.data.type) return next(new Error('JSON API resource object must contain `data.type` property'));
-      if (ctx.req.method === 'PATCH') {
-        if (!data.data.id) return next(new Error('JSON API resource object must contain `data.id` property'));
+      errors = validateRequest(data, ctx.req.method);
```
Comment (digitalsadhu) [inline lib/deserialize.js]: vast improvement

File: lib/deserializer.js [hunk 6]
```diff
@@ -1,57 +1,78 @@
...
-        if (serverRelation.foreignKey) {
-          fk = serverRelation.foreignKey;
-        } else {
-          relationType = inflection.camelize(relationType, true);
-          fk = relationType + fkSuffix;
-        }
-        result[fk] = relationValue;
-      }
+    _.each(clientRelations, function (relation, name) {
+      handleClientRelations(data, result, serverRelations[name], relation);
```
Comment (digitalsadhu) [inline lib/deserializer.js]: nice to split that out, cleans it up nicely

File: lib/headers.js (hunk 1 above)
Comment (digitalsadhu) [inline lib/headers.js]: ah, perhaps based on reuse?

File: lib/removeRemoteMethods.js [hunk 7]
```diff
@@ -4,12 +4,10 @@ module.exports = function (app, options) {
...
   var isStatic = true;
+
   Object.keys(models).forEach(function (model) {
-    models[model].disableRemoteMethod('upsert', isStatic);
-    models[model].disableRemoteMethod('exists', isStatic);
-    models[model].disableRemoteMethod('findOne', isStatic);
-    models[model].disableRemoteMethod('count', isStatic);
-    models[model].disableRemoteMethod('createChangeStream', isStatic);
-    models[model].disableRemoteMethod('updateAll', isStatic);
+    ['upsert', 'exists', 'findOne', 'count', 'createChangeStream', 'updateAll'].forEach(function (method) {
```
Comment (digitalsadhu) [inline lib/removeRemoteMethods.js]: :+1:

File: package.json [hunk 8]
```diff
@@ -5,6 +5,7 @@
   "main": "lib/index.js",
   "scripts": {
     "test": "npm run lint && mocha --reporter=spec ./test/**/*.test.js",
+    "tester": "mocha --reporter=spec ./test/**/*.test.js",
```
Comment (digitalsadhu) [inline package.json]: just so you didn't have to run linting every time you ran tests?

**Comment by digitalsadhu** [comment]
Im going to merge and get on with trying to fix that breaking test. Be good to get back to green. ;)

**Comment by digitalsadhu** [comment]
@tsteuwer 
Actually we'll need to get the code style failures cleaned up first.

FYI I think it best if we use the same style rules as strongloop uses in case we ever end up moving this component into core. I'm not actually sure how closely our style rules align with strongloops atm. Would be worth checking. doesn't have to be now.

For now we should either fix to the current rules or change the rules so they things pass.

**Comment by tsteuwer** [comment]
Sweet :)

**Comment by tsteuwer** [comment]
When I ran them there were no errors?

**Comment by digitalsadhu** [comment]
Could you have run npm run tester vs npm test? 
I haven't looked into it beyond failures on CI and ran locally to confirm and got the same result.

**Comment by digitalsadhu** [comment]
@tsteuwer was mostly tabs vs spaces. Did a quick format to the current standard as thought that was easiest.

**Comment by digitalsadhu** [comment]
ok merging

**Pull request merged**